### DISCUSSION TITLE
Help me understand paging chapters

### DISCUSSION THREAD

**Discussion by gyzerok** [phil-opp/blog_os#1331] [Questions]
Hello there!

Even though I come here with some struggles and questions let me start by saying that the whole guide is awesome! While I can't quite comprehend everything written, I feel like I've expanded my knowledge dramatically while reading it.

Now let's get back to paging.

In the beginning of the article identity mappings are described and how they are cluttering the virtual address space. What I don't understand here is why is that? What is cluttering here? Why can't we allocate enough page tables (to map entire physical memory) at the beginning of the virtual address space and then use the rest of the address space for the actual usable memory?

For example let's say we have about 2mb of physical memory. Then we need just a single table (4kb) to map it. We can use the 0x1 frame to store that page and then use addresses from 0x2 for the usable memory. This addresses are continuous. We can scale it to several tables if we have more physical memory, but still place them in the first several addresses. What am I missing?

When we later talk about fixed offset mapping it seems to me exactly like my imaginary example above with the only difference that we place page table(s) not in the beginning but rather "in the end", so we could use virtual addressses at the start for "usable memory". What am I missing again?

Later on we enable memory mapping in the bootloader and it is said that bootloader will map complete physical memory for us. But if bootloader already mapped complete physical memory then why do we create frame allocator to map pages? Aren't they all mapped already?

I feel like I have even more questions, but I even struggle to articulate them. Hopefully answers to these ones will help me figure something out and ask further. Please help me figure everything out :)

**Comment by geeky-ashu** [comment]
Your intuitions are largely right - let me clarify each part.

Why does identity mapping clutter virtual address space?

With identity mapping, each physical frame used for a page table occupies the same virtual address. If you want a 1 MiB contiguous virtual memory region (for a file mapping or large buffer), but scattered page table frames sit at physical addresses 200 KiB, 600 KiB, and 900 KiB, those virtual addresses are already taken. You can't span a contiguous region across any of them - you have to hunt for a gap large enough. That fragmentation is what the blog means by "cluttering."

Why can't you just put page tables at the start of virtual address space? And isn't fixed-offset mapping the same thing?
You are correct on both counts. Fixed-offset mapping is exactly your idea - the difference is only where the offset is placed: at a high virtual address instead of starting at 0.

The reason for using a high offset is that user-space programs expect their virtual address space to start near 0. A process has its executable near the bottom, heap growing upward, stack near the top of its range. If the kernel claimed the first few gigabytes at low addresses for its physical memory mapping, user processes would have nowhere to run. The x86_64 convention is: lower half (below 0x8000_0000_0000) for user space, upper half for kernel.

Putting the mapping at low addresses would work for a kernel-only system, but breaks as soon as you introduce user processes.

If the bootloader maps all physical memory, why do we still need a frame allocator?

These are two different things:
- Mapping = creating page table entries so the kernel can access any byte of RAM. Think of it as giving the kernel a master key that opens every room in a building - it can read and write anything. But having a key doesn't mean the rooms are empty.
- Frame allocating = tracking which physical frames are free vs. already in use. The kernel code, stack, boot info, and firmware structures all occupy specific physical frames. When you need a new frame - to back a new page table level or a heap allocation - you ask the frame allocator "which frames are unused?" so you don't overwrite something that's already there.

The BootInfoFrameAllocator reads the memory map the bootloader provides to find frames marked Usable (not reserved for kernel, firmware, or MMIO). Without it, there's no way to safely get a new frame to write a fresh page table into.
